runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/Deserializer.kt

```diff
@@ -198,7 +198,7 @@ interface PrimitiveDeserializer {
      * If the document's value is a list or map, this method will deserialize all elements or fields recursively - the
      * caller need not further inspect the value to attempt to do so manually.
      */
-    fun deserializeDocument(): Document
+    fun deserializeDocument(): Document?
```

Comment:
I don't _think_ we need the `?` qualifier. An explicit `null` is the same as not sending it in the payload as far as serialization and deserialization is concerned. JSON for example [won't even report `null` values](https://github.com/awslabs/smithy-kotlin/blob/2547699c13857f5dbd57447ff5e2bdf0308aeefa/runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonDeserializer.kt#L195).

Reply:
Hmm I guess maybe this is due to the internal nullability of a document? In which case you _could_ make the deserializer interface non-nullable return and delegate intrernally to a function that allows it to be `null`

Reply:
Can you clarify your objection to it being nullable? It's done that way because of internal Document nullability as you suspected.

It certainly can be split out like you said, but it'll just bulk up and duplicate some code (you'll basically have this deserialize switch twice), and to avoid what I'm not quite sure.

Reply:
It's different than how the rest of the deserializer interface works. 

We only ever call `deserialize<Type>` when we have found a matching field:


```kotlin
deserializer.deserializeStruct(OBJ_DESCRIPTOR) {
    loop@while (true) {
        when (findNextFieldIndex()) {
            MY_DOCUMENT_MEMBER.index -> builder.myDocumentMember = deserializeDocument()
            ...
        }
    }
}
```

`deserializeDocument()` will only ever be called if the value associated with the key represented by `MY_DOCUMENT_MEMBER` is non-null. e.g.

```json
{
     "myDocumentMember": "foo"
}
```

whereas: 
```json
{
     "myDocumentMember":  null
}
```

will never execute that branch.


It's maybe a minor thing overall I'm just questioning whether we should codify that difference. Internally a document may be null so I would expect _if_ we wanted the top level deserializer function to return a non-null document it could delegate like so:

```
fun deserializeDocument(): Document = checkNotNull(deserializeDocumentImpl()) { "expected non-null document field" }

private fun deserializeDocumentImpl(): Document? { ... what you have now basically }
```

Reply:
As long as our existing deserializer logic correctly shortcuts explicitly-null member and absent member (i.e., implicitly-null) then I favor using the same semantics as other methods. That is, `deserializeDocument(): Document` (non-null).